src/mococrw/dilithium.h

```diff
@@ -0,0 +1,378 @@
+     */
+    static DilithiumAsymmetricKeypair readPrivateKeyfromDER(const std::vector<uint8_t> &asn1Data);
+
+    /**
+     * Generate an asymmetric keypair with given Spec.
+     *
+     * @see DilithiumSpec
+     *
+     * @throws This method may throw an MoCOCrWException if dilithium
+     *      indicates an error or an invalid parameter set is used.
+     */
+    static DilithiumAsymmetricKeypair generate(const DilithiumSpec &spec);
+
+private:
+    DilithiumAsymmetricKeypair(DilithiumAsymmetricKey &&key)
+            : DilithiumAsymmetricPublicKey{std::move(key)}
+    {
+    }
+};
+
+/**
+ * For readability, this alias may sometime prove useful.
+ */
+using DilithiumAsymmetricPrivateKey = DilithiumAsymmetricKeypair;
+
+/**
+ * @brief Specification for Dilithium keys
+ *
+ * This class allows to generate Dilithium keypairs.
+ * It shall inherit from AsymmetricKey class once openssl supports dilithium.
+ * To support this the class implements the same functions as required by
+ * AsymmetricKey.
+ *
+ */
+class DilithiumSpec
+{
+public:
+    /* According to https://pq-crystals.org/dilithium/ DILITHIUM3 shall be used
+     * as default parameter set.
+     */
+    static const DilithiumKeyImpl::DilithiumParameterSet defaultDilithiumParamSet =
+            DilithiumKeyImpl::DilithiumParameterSet::DILITHIUM3;
+
+    explicit DilithiumSpec(DilithiumKeyImpl::DilithiumParameterSet paramSet) : _paramSet{paramSet}
+    {
+    }
+
+    DilithiumSpec() : DilithiumSpec{defaultDilithiumParamSet} {}
+    DilithiumAsymmetricKey generate() const;
+
+private:
+    DilithiumKeyImpl::DilithiumParameterSet _paramSet;
+};
+
+class DilithiumSigningCtx : public MessageSignatureCtx
+{
+public:
+    DilithiumSigningCtx(const DilithiumAsymmetricPrivateKey &key);
+
+    ~DilithiumSigningCtx();
+    /**
+     * @brief Copy Constructor
+     */
+    DilithiumSigningCtx(const DilithiumSigningCtx &other);
+
+    /**
+     * @brief Copy Assignment
+     */
+    DilithiumSigningCtx &operator=(const DilithiumSigningCtx &other);
+
+    std::vector<uint8_t> signMessage(const std::vector<uint8_t> &message);
+
+private:
+    /**
+     * Internal class for applying the PIMPL design pattern
+     * (to hide the details of storing the padding objects from the client)
+     */
+    class Impl;
+
+    /**
+     * Pointer for PIMPL design pattern
+     */
+    std::unique_ptr<Impl> _impl;
+};
+
+class DilithiumVerificationCtx : public MessageVerificationCtx
+{
+public:
+    DilithiumVerificationCtx(const DilithiumAsymmetricPublicKey &key);
+    ~DilithiumVerificationCtx();
+    /**
+     * @brief Copy Constructor
+     */
+    DilithiumVerificationCtx(const DilithiumVerificationCtx &other);
+
+    /**
+     * @brief Copy Assignment
+     */
+    DilithiumVerificationCtx &operator=(const DilithiumVerificationCtx &other);
+
+    void verifyMessage(const std::vector<uint8_t> &signature, const std::vector<uint8_t> &message);
```

Comment:
hm, no `verifyDigest`?

I'll guess then that `verifyMessage` does not do any internal `digest` with a known hashing algorithm (e.g., sha256). If this is not true, it could be a problem for applications

Reply:
see above.